### ISSUE TITLE
Does the  armnn support  subgraph  for tensorflow now?

### ISSUE THREAD

**Issue by Cola-Code** [ARM-software/armnn#142]
I tried to deploy a mobilefacenet-tensorflow model with 3 subgraphs but found that the code cannot return from IRuntime::LoadNetwork.Did this cause by the subgraphs?Any suggestions for this problem?

**Comment by MatthewARM** [comment]
Hi @Cola-Code I was not aware that Tensorflow had subgraphs, so that could indeed be causing problems.

I know that Tensorflow Lite has subgraphs,  if you are using the ArmNN Tensorflow Lite parser then that will produce an error. When we wrote that parser we hadn't seen a .tflite file with more than one subgraph.

Can you link to the model you are using? Then we can look into how difficult it will be to support it.

**Comment by Cola-Code** [comment]
Hi @MatthewARM,
Sorry i made a mistake. I was confused by the scope and subgraph.
And  the real reason for the problem is that I replaced the ELU with the combine and relu and sigmoid:

def my_elu_sigmoid(input):
    output = tf.nn.relu(input) + (tf.div(1.0, tf.nn.sigmoid(tf.nn.relu(input) - input)) - 2)
    return output

The replacement caused many new branchs in model, and increased the complexity of the model. In LayerSelectionInfo::MarkChildrenSplits, the code would traval to the end for each branch. So the time consuming increased exponentially nearly after replacement.
At last, I add the ELU ops in armnn and solved the problem.

I was doubt that it's nesscessary to traval to  the last ops in the model in MarkChildrenSplits? Or i understood wrongly?

**Comment by MatthewARM** [comment]
Hi @Cola-Code very interesting, thanks.

Two questions:
* Which version of armnn are you using? 18.11 or master?
* Did the new network have cycles? I can see that MarkChildrenSplits doesn't check for cycles, so if you somehow managed to get a cyclic graph into ArmNN then it would not return.

The other possibility is that it simply ground to a halt due to excessive recursion and running low on stack space.  It certainly seems like we will need to look at this area in order to improve handling very large models.

Have you made code changes related to ELU? Can I encourage you to submit them to Gerrit at https://review.mlplatform.org/#/admin/projects/ml/armnn ?

**Comment by Cola-Code** [comment]
I download the armnn-master code from https://git.mlplatform.org/ml/armnn.git on Feb.12.

For finding out the problem, i added some debug message in code:

void MarkChildrenSplits(LayerInfoContainer& network,
                            uint32_t splitId,
                            bool prevSelected)
    {
        /*static int tmp_id = 0;
        for (int i = 0; i < tmp_id; i++)
        {
            std::cout << " ";
        }
        std::cout << "L-" << tmp_id << ": " << m_Layer->GetNameStr() << std::endl;*/
        if (m_SplitId < splitId)
        {
            m_SplitId = splitId;
        }

        // introduce a new split point at all non-selected points, but only if the
        // previous point was selected. this prevents creating a new subgraph at
        // every non-selected layer
        if (!m_IsSelected && prevSelected)
        {
            ++m_SplitId;
        }
        
        for (auto& layer : m_DirectChildren)
        {
            auto it = network.find(layer);
            //tmp_id++;
            BOOST_ASSERT_MSG(it != network.end(), "All layers must be part of the topology.");
            if (it != network.end())
            {
                it->second.MarkChildrenSplits(network, m_SplitId, m_IsSelected);
            }
            //tmp_id--;
        }
    }

**Comment by Cola-Code** [comment]
I download the armnn-master code from https://git.mlplatform.org/ml/armnn.git on Feb.12.

For finding out the problem, i added some debug message in code:

void MarkChildrenSplits(LayerInfoContainer& network,
                            uint32_t splitId,
                            bool prevSelected)
    {
        /*static int tmp_id = 0;
        for (int i = 0; i < tmp_id; i++)
        {
            std::cout << " ";
        }
        std::cout << "L-" << tmp_id << ": " << m_Layer->GetNameStr() << std::endl;*/
        if (m_SplitId < splitId)
        {
            m_SplitId = splitId;
        }

        // introduce a new split point at all non-selected points, but only if the
        // previous point was selected. this prevents creating a new subgraph at
        // every non-selected layer
        if (!m_IsSelected && prevSelected)
        {
            ++m_SplitId;
        }
        
        for (auto& layer : m_DirectChildren)
        {
            auto it = network.find(layer);
            //tmp_id++;
            BOOST_ASSERT_MSG(it != network.end(), "All layers must be part of the topology.");
            if (it != network.end())
            {
                it->second.MarkChildrenSplits(network, m_SplitId, m_IsSelected);
            }
            //tmp_id--;
        }
    }

**Comment by Cola-Code** [comment]
I download the armnn-master code from https://git.mlplatform.org/ml/armnn.git on Feb.12.

For finding out the problem, i added some debug message in code:

void MarkChildrenSplits(LayerInfoContainer& network,
                            uint32_t splitId,
                            bool prevSelected)
    {
        /*static int tmp_id = 0;
        for (int i = 0; i < tmp_id; i++)
        {
            std::cout << " ";
        }
        std::cout << "L-" << tmp_id << ": " << m_Layer->GetNameStr() << std::endl;*/
        if (m_SplitId < splitId)
        {
            m_SplitId = splitId;
        }

        // introduce a new split point at all non-selected points, but only if the
        // previous point was selected. this prevents creating a new subgraph at
        // every non-selected layer
        if (!m_IsSelected && prevSelected)
        {
            ++m_SplitId;
        }
        
        for (auto& layer : m_DirectChildren)
        {
            auto it = network.find(layer);
            //tmp_id++;
            BOOST_ASSERT_MSG(it != network.end(), "All layers must be part of the topology.");
            if (it != network.end())
            {
                it->second.MarkChildrenSplits(network, m_SplitId, m_IsSelected);
            }
            //tmp_id--;
        }
    }

**Comment by Cola-Code** [comment]
I download the armnn-master code from https://git.mlplatform.org/ml/armnn.git on Feb.12.

For finding out the problem, i added some debug message in code:

void MarkChildrenSplits(LayerInfoContainer& network,
                            uint32_t splitId,
                            bool prevSelected)
    {
        /*static int tmp_id = 0;
        for (int i = 0; i < tmp_id; i++)
        {
            std::cout << " ";
        }
        std::cout << "L-" << tmp_id << ": " << m_Layer->GetNameStr() << std::endl;*/
        if (m_SplitId < splitId)
        {
            m_SplitId = splitId;
        }

        // introduce a new split point at all non-selected points, but only if the
        // previous point was selected. this prevents creating a new subgraph at
        // every non-selected layer
        if (!m_IsSelected && prevSelected)
        {
            ++m_SplitId;
        }
        
        for (auto& layer : m_DirectChildren)
        {
            auto it = network.find(layer);
            //tmp_id++;
            BOOST_ASSERT_MSG(it != network.end(), "All layers must be part of the topology.");
            if (it != network.end())
            {
                it->second.MarkChildrenSplits(network, m_SplitId, m_IsSelected);
            }
            //tmp_id--;
        }
    }

**Comment by Cola-Code** [comment]
I download the armnn-master code from https://git.mlplatform.org/ml/armnn.git on Feb.12.

For finding out the problem, i added some debug message in code:

void MarkChildrenSplits(LayerInfoContainer& network,
                            uint32_t splitId,
                            bool prevSelected)
    {
        /*static int tmp_id = 0;
        for (int i = 0; i < tmp_id; i++)
        {
            std::cout << " ";
        }
        std::cout << "L-" << tmp_id << ": " << m_Layer->GetNameStr() << std::endl;*/
        if (m_SplitId < splitId)
        {
            m_SplitId = splitId;
        }

        // introduce a new split point at all non-selected points, but only if the
        // previous point was selected. this prevents creating a new subgraph at
        // every non-selected layer
        if (!m_IsSelected && prevSelected)
        {
            ++m_SplitId;
        }
        
        for (auto& layer : m_DirectChildren)
        {
            auto it = network.find(layer);
            //tmp_id++;
            BOOST_ASSERT_MSG(it != network.end(), "All layers must be part of the topology.");
            if (it != network.end())
            {
                it->second.MarkChildrenSplits(network, m_SplitId, m_IsSelected);
            }
            //tmp_id--;
        }
    }

**Comment by Cola-Code** [comment]
I download the armnn-master code from https://git.mlplatform.org/ml/armnn.git on Feb.12.

For finding out the problem, i added some debug message in code:

void MarkChildrenSplits(LayerInfoContainer& network,
                            uint32_t splitId,
                            bool prevSelected)
    {
        static int tmp_id = 0;
        for (int i = 0; i < tmp_id; i++)
        {
            std::cout << " ";
        }
        std::cout << "L-" << tmp_id << ": " << m_Layer->GetNameStr() << std::endl;
        if (m_SplitId < splitId)
        {
            m_SplitId = splitId;
        }

        // introduce a new split point at all non-selected points, but only if the
        // previous point was selected. this prevents creating a new subgraph at
        // every non-selected layer
        if (!m_IsSelected && prevSelected)
        {
            ++m_SplitId;
        }
        
        for (auto& layer : m_DirectChildren)
        {
            auto it = network.find(layer);
            tmp_id++;
            BOOST_ASSERT_MSG(it != network.end(), "All layers must be part of the topology.");
            if (it != network.end())
            {
                it->second.MarkChildrenSplits(network, m_SplitId, m_IsSelected);
            }
            tmp_id--;
        }
    }

And I got the message like this:
L-86: MobileFaceNet/MobileFaceNet/Conv2d_5_InvResBlock_0/Conv_1/BatchNorm/FusedBatchNorm
                                                                                       L-87: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/Conv/Conv2D
                                                                                        L-88: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/Conv/BatchNorm/FusedBatchNorm
                                                                                         L-89: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/Conv/Elu
                                                                                          L-90: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/SeparableConv2d/depthwise
                                                                                           L-91: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/SeparableConv2d/BatchNorm/FusedBatchNorm
                                                                                            L-92: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/SeparableConv2d/Elu
                                                                                             L-93: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/Conv_1/Conv2D
                                                                                              L-94: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/Conv_1/BatchNorm/FusedBatchNorm
                                                                                               L-95: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/Add
                                                                                                L-96: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv/Conv2D
                                                                                                 L-97: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv/BatchNorm/FusedBatchNorm
                                                                                                  L-98: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv/Elu
                                                                                                   L-99: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/SeparableConv2d/depthwise
                                                                                                    L-100: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/SeparableConv2d/BatchNorm/FusedBatchNorm
                                                                                                     L-101: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/SeparableConv2d/Elu
                                                                                                      L-102: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv_1/Conv2D
                                                                                                       L-103: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv_1/BatchNorm/FusedBatchNorm
                                                                                                        L-104: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Add
                                                                                                         L-105: MobileFaceNet/MobileFaceNet/Conv2d_7/Conv2D
                                                                                                          L-106: MobileFaceNet/MobileFaceNet/Conv2d_7/BatchNorm/FusedBatchNorm
                                                                                                           L-107: MobileFaceNet/MobileFaceNet/Conv2d_7/Elu
                                                                                                            L-108: MobileFaceNet/Logits/SeparableConv2d/depthwise
                                                                                                             L-109: MobileFaceNet/Logits/SeparableConv2d/BatchNorm/FusedBatchNorm
                                                                                                              L-110: MobileFaceNet/Logits/Conv/Conv2D
                                                                                                               L-111: MobileFaceNet/Logits/Conv/BatchNorm/FusedBatchNorm
                                                                                                                L-112: MobileFaceNet/Logits/LinearConv1x1/Conv2D
                                                                                                                 L-113: MobileFaceNet/Logits/LinearConv1x1/BatchNorm/FusedBatchNorm
                                                                                                                  L-114: MobileFaceNet/Logits/SpatialSqueeze
                                                                                                                   L-115: MobileFaceNet/Logits/SpatialSqueeze
                                                                                                L-96: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Add
                                                                                                 L-97: MobileFaceNet/MobileFaceNet/Conv2d_7/Conv2D
                                                                                                  L-98: MobileFaceNet/MobileFaceNet/Conv2d_7/BatchNorm/FusedBatchNorm
                                                                                                   L-99: MobileFaceNet/MobileFaceNet/Conv2d_7/Elu
                                                                                                    L-100: MobileFaceNet/Logits/SeparableConv2d/depthwise
                                                                                                     L-101: MobileFaceNet/Logits/SeparableConv2d/BatchNorm/FusedBatchNorm
                                                                                                      L-102: MobileFaceNet/Logits/Conv/Conv2D
                                                                                                       L-103: MobileFaceNet/Logits/Conv/BatchNorm/FusedBatchNorm
                                                                                                        L-104: MobileFaceNet/Logits/LinearConv1x1/Conv2D
                                                                                                         L-105: MobileFaceNet/Logits/LinearConv1x1/BatchNorm/FusedBatchNorm
                                                                                                          L-106: MobileFaceNet/Logits/SpatialSqueeze
                                                                                                           L-107: MobileFaceNet/Logits/SpatialSqueeze
                                                                                       L-87: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_0/Add
                                                                                        L-88: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv/Conv2D
                                                                                         L-89: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv/BatchNorm/FusedBatchNorm
                                                                                          L-90: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv/Elu
                                                                                           L-91: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/SeparableConv2d/depthwise
                                                                                            L-92: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/SeparableConv2d/BatchNorm/FusedBatchNorm
                                                                                             L-93: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/SeparableConv2d/Elu
                                                                                              L-94: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv_1/Conv2D
                                                                                               L-95: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Conv_1/BatchNorm/FusedBatchNorm
                                                                                                L-96: MobileFaceNet/MobileFaceNet/Conv2d_6_InvResBlock_1/Add
                                                                                                 L-97: MobileFaceNet/MobileFaceNet/Conv2d_7/Conv2D
                                                                                                  L-98: MobileFaceNet/MobileFaceNet/Conv2d_7/BatchNorm/FusedBatchNorm
                                                                                                   L-99: MobileFaceNet/MobileFaceNet/Conv2d_7/Elu
                                                                                                    L-100: MobileFaceNet/Logits/SeparableConv2d/depthwise
                                                                                                     L-101: MobileFaceNet/Logits/SeparableConv2d/BatchNorm/FusedBatchNorm
                                                                                                      L-102: MobileFaceNet/Logits/Conv/Conv2D
                                                                                                       L-103: MobileFaceNet/Logits/Conv/BatchNorm/FusedBatchNorm
                                                                                                        L-104: MobileFaceNet/Logits/LinearConv1x1/Conv2D
                                                                                                         L-105: MobileFaceNet/Logits/LinearConv1x1/BatchNorm/FusedBatchNorm
                                                                                                          L-106: MobileFaceNet/Logits/SpatialSqueeze
                                                                                                           L-107: MobileFaceNet/Logits/SpatialSqueeze

MobileFaceNet/Logits/SpatialSqueeze is the last node in the graph.

**Comment by MatthewARM** [comment]
Oh dear it looks like MarkChildrenSplits has a problem when there is a join in graph, e.g. when there is an Add where both operands of the Add come from branches in the graph. In these cases, the Add and everything after it is processed more than once.

We should be able to fix this fairly easily but not in time for the 19.02 release.

**Comment by MatthewARM** [comment]
Hi @Cola-Code the fix for this issue is in master here: https://review.mlplatform.org/#/c/ml/armnn/+/805/

Unfortunately it was not possible to get the fix ready in time for 19.02, but the patch should apply cleanly to 19.02 for you to evaluate.

All the best,
Matthew

**Comment by MatthewARM** [comment]
This was fixed in 19.05
